web-animations/timing-model/animations/updating-the-finished-state.html

```diff
@@ -0,0 +1,131 @@
+  t.step(function() {
+    assert_equals(event.target, animation, "Animation should be event target");
+    assert_times_equal(event.currentTime, expectedEventTime,
+        "Event currentTime should be the given expected value");
+    assert_times_equal(event.timelineTime, animation.startTime + expectedEventTime,
+        "Event timelineTime should be the animation's start time plus the given expected value");
+  });
+}
+
+function awaitFinishEventAndPromise(t, animation, expectedEventTime) {
+  var eventPromise = new Promise(function(resolve) {
+    animation.onfinish = function(event) {
+      validateFinishEvent(t, event, animation, expectedEventTime);
+      resolve();
+    };
+  });
+
+  var finishedPromise = animation.finished.then(function(target) {
+    assert_equals(target, animation);
+  });
+
+  var timeout = waitForAnimationFrames(3).then(function() {
+    return Promise.reject("Expected finish notifications did not occur");
+  });
+
+  return Promise.race([timeout,Promise.all([eventPromise, finishedPromise])]);
+}
+
+promise_test(function(t) {
+  var div = createDiv(t);
+  var animation = div.animate(null, 1);
+  animation.onfinish = t.unreached_func("Seeking to finish should not fire finish event");
+  animation.finished.then(t.unreached_func("Seeking to finish should not resolve finished promise"));
+  animation.currentTime = 1;
+  animation.currentTime = 0;
+  animation.pause();
+  return waitForAnimationFrames(3);
+}, "Finish notification steps don't run when the animation seeks to finish and then seeks back again");
+
+promise_test(function(t) {
+  var div = createDiv(t);
+  var animation = div.animate(null, 1);
+  animation.onfinish = t.unreached_func("Seeking past finish should not fire finish event");
+  animation.finished.then(t.unreached_func("Seeking past finish should not resolve finished promise"));
+  animation.currentTime = 10;
+  animation.currentTime = 0;
+  animation.pause();
+  return waitForAnimationFrames(3);
+}, "Finish notification steps don't run when the animation seeks past finish and then seeks back again");
+
+promise_test(function(t) {
+  var div = createDiv(t);
+  var animation = div.animate(null, 1);
+  animation.finish();
+  animation.currentTime = 0;
+  animation.pause();
+  var expectedEventTime = 1;
+  return awaitFinishEventAndPromise(t, animation, expectedEventTime);
+}, "Finish notification steps run when the animation completes with .finish(), even if we then seek away");
+
+promise_test(function(t) {
+  var div = createDiv(t);
+  var animation = div.animate(null, 1);
+  animation.currentTime = 10;
+  var expectedEventTime = 10;
+  return awaitFinishEventAndPromise(t, animation, expectedEventTime);
+}, "Finish notification steps run when the animation seeks past finish");
+
+promise_test(function(t) {
+  var div = createDiv(t);
+  var animation = div.animate(null, 1);
+  var expectedEventTime = 1;
+  return awaitFinishEventAndPromise(t, animation, expectedEventTime);
```

Comment:
Oddly enough this takes three animation frames in Firefox so we fail the timeout here. I think what happens is we take two frames to start the animation, we resolve the ready promise and finished promise in the same frame, but we schedule the dispatch of the finish event asynchronously and on the next frame, the rAF callback runs before the event handler does so we just miss the timeout.

Can we wait for `animation.ready` first?